Tagcloud

thomasneirynck opened the pull request:
See #8149 

moving tagcloud PR (#6975). resolved merge conflicts, checking if it runs in CI

thomasneirynck commented:
jenkins, test this.

thomasneirynck commented:
This is now running again. I added it to the core-vis library, so it loads on startup without additional plugin installation commands.

(Don't forget to run `npm install` !!)

This still needs:
- [ ] evaluate all functionality
- [ ] verify/extend existing test suite
- [ ] bring up to date with most recent code-style
- [ ] settle on appropriate place for this (core, extra plugin, ...)

markharwood commented:
If using significant terms is there an option to size text by the significance score?
Popularity != significance e.g. the circle sizes [in this example visualization](https://www.quora.com/Whats-the-best-way-to-display-significant-terms-aggregation-in-Kibana) are sized by level-of-change not popularity

stormpython commented on src/core_plugins/tag_cloud/public/lib/cloud_controller.js:
```diff
@@ -0,0 +1,34 @@
+    const tags = buckets.map(function (bucket) {
+      return {
+        text: bucket.key,
+        size: metricsAgg.getValue(bucket)
```
@markharwood @thomasneirynck 
here is where the size for the text is set.

thomasneirynck commented:
@markharwood not sure. Have you seen this in any other Kibana visualizations? The tags are sized according to the aggregation-value. So e.g. counts, averages, ... I'm not quite sure if we are lifting out the significance score in any other viz either.

markharwood commented:
>here is where the size for the text is set.

I'm not sure what metricsAgg is picking up as the value - the score or the doc count?
If your search is for `bird flu` doc count will tell you that `the` is most popular while significance score will tell you `h5n1` is most interesting.

> I'm not quite sure if we are lifting out the significance score in any other viz either.

We use it as the sort order for selecting which terms to surface but don't illustrate the magnitude of the score. This is what the diagram I linked to illustrates in terms of movement from a background frequency. It's possibly a little hard for users to interpret but the sizing of words in tag clouds is easier to grok and should be based on score not doc_count.

stormpython commented:
@markharwood @thomasneirynck its picking up the doc count. If I recall correctly, `metricsAgg` is from the Metrics dropdown that is selected, which defaults to `count`. I am not sure if significance score is an option in the Metrics dropdown. But if its added for significant terms, then it should just work out of the box.

thomasneirynck commented:
for reviewers, this is an early check-in, so this will be revised most certainly.

For now, mainly looking for functional feedback: do the options in the forms make sense, does it work as expected, what is missing/what should be added, ...

jbudz commented:
Awesome, I like this a lot, this looks really good.  Not looking at the code but a few notes while using it:
![image](https://cloud.githubusercontent.com/assets/3143860/18443724/d0a617da-78db-11e6-8e29-2d6bfc563a99.png)
![image](https://cloud.githubusercontent.com/assets/3143860/18443795/2852da5e-78dc-11e6-97eb-db7eed8d9ba0.png)

* The current font doesn't match kibana's font family
* The arc input is handy but also it can be difficult to reset it back to defaults
* I'm not sure if it's possible with the current plugin, but word positioning isn't deterministic.  On page refresh I can see it render twice, resize, etc moving the words around

thomasneirynck commented:
@jbudz 
 
about the determinism of word positioning. Agreed, that's a problem. At  this point,I'm not quite sure how to resolve this. The problem is that at each visualization rendering, the entire  visualization gets torn down and recreated (the layout algo indeed has some random elements). This is the standard flow of the visualization app. Ideally, there's an update hook, that when the visualizations options change (and the data remains the same), the same order can be reused. To be investigated.It'd be nice to incorporate this.

(thanks for the bug list, I'll address these)

cjcenizal commented:
Great work on this @thomasneirynck ! And thanks for coming back and offering guidance @stormpython . :)

A few comments:

**Difficulty in relating input to output**

I had a bit of a hard time understanding how my aggregations were impacting the tag cloud. This is a problem I have with _all_ visualizations, but it seems a little more acute here because of the "volatile" nature of the tag cloud. This may be due to its non-determinism. It may also be due to the fact that tag clouds represent data in less-organized way, so it's more difficult for me to pin down specific relationships between different parts of the vis (as opposed to say, a bar chart).

**Font style and weight**

Maybe we can remove the oblique font style, as well as the bolder and lighter weight options. They didn't seem to have a noticeable impact on the way the vis appeared.

**Orientations control**

Maybe we can brainstorm on ways to make this control clearer? I don't quite understand how this control works. Possibly we could solve this with better labeling/help text.

**The usefulness of tag clouds**

Here are some useful articles I found on the pros and cons of tag clouds:

- http://ux.stackexchange.com/questions/8158/are-tag-word-clouds-useful
- https://www.visioncritical.com/pros-and-cons-word-clouds-visualizations/
- http://www.niemanlab.org/2011/10/word-clouds-considered-harmful/ (Very strongly against!)

Personally, I don't think they are very useful but I'm sure that we have users who would like to use them. So I think this is a good addition to our product, and I think our users will appreciate it!

thomasneirynck commented:
@cjcenizal thanks so much for the careful and measured feedback, I love that.

- orientations control: this is something I experimented with, so it's good to have other people try it out. I wanted to avoid the ability of users to put in a  conflicting feedback, and not have to deal with 3 separate controls. Mixed results, so let's talk ;)
- usefulness: I think word clouds are good as an infographic. Something that's a little whimsical, and allows presenters to make part of their story 'pop'.  It has a narrow range of usefulness. It is not good as a data visualization, in that you can only make the crudest quantitative inference about the underlying data. So I'd tend to agree with the  first two links, who balance both ends in the same way. I think the message from the 3rd link is a little too glib; never say never. It is also a critique of using too simple of an underlying textual analysis (word counting as some sort of indication  of relevance), that would reflect in any visualization using a term aggregation in Kibana.

cjcenizal commented on src/core_plugins/tag_cloud/public/arcinput_directive.js:
```diff
@@ -26,7 +26,12 @@ module.directive('arcInput', function ($timeout) {
       scope.$on('$destroy', () => {
         $timeout.cancel(resizeTimeout);
-        element[0].firstChild.children[0].removeEventListener(onOrientationChange);
+        try {//gross, but watill be deletd anyway
```
Can you elaborate on this comment?

thomasneirynck replied:
sorry, that file needs to be deleted. it was a mistake to check this in, let alone the garbled language

cjcenizal commented on src/core_plugins/tag_cloud/public/vis/components/layout/layout.js:
```diff
@@ -71,24 +74,67 @@ function baseLayout() {
 
-export default baseLayout;
+function translate(d) {
+  return 'translate(' + d.dx + ',' + d.dy + ')';
```
```javascript
return `translate(${d.dx},${d.dy})`;
```

jbudz commented:
when you get a chance can we rebase this?

thomasneirynck commented:
Latest:
- removed gridded layouts. This is only relevant when we have multiple clouds. For now, that's not necessary.
- trimmed code down, simplified directory structure
- added animations between different clouds. this helps avoid nasty flickers as well.

cjcenizal commented on src/core_plugins/tag_cloud/public/cloud.js:
```diff
@@ -0,0 +1,60 @@
+import tagCloudVisParamsTemplate from 'plugins/tagcloud/cloud_vis_params.html';
+
+
+require('ui/registry/vis_types').register(TagCloudProvider);
```
Minor nit: Maybe we shouldn't mix `import` and `require`?

cjcenizal commented on src/core_plugins/tag_cloud/public/cloud_controller.js:
```diff
@@ -0,0 +1,35 @@
+
+    $scope.data = [{tags: tags}];
+  });
+});
```
Can we leave a breadcrumb trail for future devs who come to this file? Can we document where esResponse is coming from? Can we add comments at each of the conditions explaining the intent/end result of each one?

cjcenizal commented on src/core_plugins/tag_cloud/public/cloud_directive.js:
```diff
@@ -0,0 +1,45 @@
+      data: '=',
+      options: '='
+    },
+    template: '<svg class="parent"></svg>',
```
Even though it's tiny, let's move this to a separate html file. This way, if someone wants to find all instances of the `parent` class in html files, or see where we're using `svg`, they can safely assume they only need to search `*.html`.

cjcenizal commented on src/core_plugins/tag_cloud/public/cloud_controller.js:
```diff
@@ -0,0 +1,35 @@
+      };
+    });
+
+    $scope.data = [{tags: tags}];
```
Do we need to set an array here? Can we just set `$scope.data = { tags }`?

cjcenizal commented on src/core_plugins/tag_cloud/public/cloud_directive.js:
```diff
@@ -0,0 +1,45 @@
+      const tagCloud = new TagCloud(element[0], containerSize());
+
+      function containerSize() {
+        return [element.parent().width(), element.parent().height()];
```
For explicitness can we rename this function to use a verb, and return an object?

```javascript
function getContainerSize() {
  return {
    width: element.parent().width(),
    height: element.parent().height(),
  }
}
```

cjcenizal commented on src/core_plugins/tag_cloud/public/tag_cloud.js:
```diff
@@ -0,0 +1,250 @@
+  'single': (tag) => 0,
+  'rightAngled': (tag) =>~~(Math.random() * 2) * 90,
+  'multi': (tag) => (~~(Math.random() * 12) * 15) - 90
+};
```
Minor formatting improvements:

```javascript
const ORIENTATIONS = {
   single: () => 0,
   rightAngled: () => ~~(Math.random() * 2) * 90,
   multi: () => (~~(Math.random() * 12) * 15) - 90,
 };
```

It would also be nice to extract these into their own module and add unit tests documenting their behavior.

thomasneirynck replied:
+1 on formatting.

cjcenizal commented on src/core_plugins/tag_cloud/public/tag_cloud.js:
```diff
@@ -0,0 +1,250 @@
+  }
+
+  setOptions(options) {
+    if (JSON.stringify(options) === this._optionBackup) {
```
I think this could use a comment explaining the intention. What's `_optionBackup`?

cjcenizal commented on src/core_plugins/tag_cloud/public/tag_cloud.js:
```diff
@@ -0,0 +1,250 @@
+        //fire rendered event here..
+      }
+    };
+    exitTransition.each(_ => exits += 1);
```
I'm not absolutely sure but can this just be:

```javascript
exits += exitTransition.nodes().length;
```

Or at the very least we can remove the `_` argument and use an increment operator:

```javascript
exitTransition.each(() => exits++);
```

cjcenizal commented on src/core_plugins/tag_cloud/public/tag_cloud.js:
```diff
@@ -0,0 +1,250 @@
+    const resolveWhenDone = () => {
+      if (exits === 0 && moves === 0) {
+        resolve(true);
+        //fire rendered event here..
```
Is this a comment a TODO? Or is there an event being fired here?

thomasneirynck replied:
this is an outstanding issue. To integrate with Reporting, we need to fire an event. Still chewing on the best way to do this.

cjcenizal commented on src/core_plugins/tag_cloud/public/tag_cloud.js:
```diff
@@ -0,0 +1,250 @@
+    this._timeoutHandle = setTimeout(() => {
+      this._timeoutHandle = null;
+      this._updateLayout();
+      /*.then(() => //..done!);*/
```
Can we delete this commented-out code?

cjcenizal commented on src/core_plugins/tag_cloud/public/tag_cloud.js:
```diff
@@ -0,0 +1,250 @@
+    tagCloudLayoutGenerator.fontStyle(this._fontStyle);
+    tagCloudLayoutGenerator.fontWeight(this._fontWeight);
+    tagCloudLayoutGenerator.fontSize(tag => this._mapSizeToFontSize(tag.size));
+    tagCloudLayoutGenerator.random(_ => 0.5); //consistently seed the layout
```
Why pass in an underscore as an argument here?

cjcenizal commented on src/core_plugins/tag_cloud/public/tag_cloud.js:
```diff
@@ -0,0 +1,250 @@
+  return {
+    size: word.size,
+    text: word.text
+  };
```
Minor suggestion but this can be terser with destructuring:

```javascript
function toWordTag(word) {
  return { size, text } = word;
}
```

cjcenizal commented on src/core_plugins/tag_cloud/public/tag_cloud.js:
```diff
@@ -0,0 +1,250 @@
+}
+
+function getSizeInPixels(tag) {
+  return tag.size + 'px';
```
```javascript
return `${tag.size}px`;
```

thomasneirynck commented:
wrt. removing font-style. That's actually a bug. I need to double check if that same bug also affects other properties (e.g. font-size).

cjcenizal commented:
@thomasneirynck Can you explain how the way we listen to resize events affects the `getContainerSize` method? I ask because I really think that using an object with human-readable width and height properties will significantly improve the readability of the code.

jbudz commented:
If there isn't enough room to fit all the words, changes/refreshing causes an update:
![tagcloud mov](https://cloud.githubusercontent.com/assets/3143860/19814989/4f866d36-9d06-11e6-8254-392220d7c00f.gif)

jbudz commented:
Custom label doesn't seem to be used, ignore the overlapping text it was mid animation:
![image](https://cloud.githubusercontent.com/assets/3143860/19815457/946c9f9a-9d08-11e6-9626-4b0ff31dc106.png)

jbudz commented:
Thoughts on adding a no results found message?:
![image](https://cloud.githubusercontent.com/assets/3143860/19815522/eab9718e-9d08-11e6-9f5c-e9b9cda7dcfa.png)

jbudz commented:
If the height is less than ~200px a scrollbar shows:
![image](https://cloud.githubusercontent.com/assets/3143860/19815839/5124350c-9d0a-11e6-85f2-54806ac28f53.png)

thomasneirynck commented:
@cjcenizal it uses angular's .watch() mechanism. Angular manages the polling. It does a deep-compare of a data-structure returned by whatever it is you are watching. In this case a 2-element array of the width/height of the container.

Oki, will change it to width/height.

ppisljar commented:
- There seem to be some performance issues when you try to render tag cloud with many tags. How do you feel about limiting the maximum number of tags you can select? Either in the aggregation itself or tag cloud simply ignoring some results from ES ?

- Another thing which I don't think is really significant: Under the metric you can still define custom label but that's not really used anywhere. Generally i think our approach in Visualize is to hide options which are not applicable. 

- sorting in ascending order on any metric but Count doesn't work (all tags disappear) and if i sort by Term then the descending mode doesn't work.

- it doesn't make sense to allow aggregating on date fields (especially if we don't use date formatter to display those fields ... right now they just render as a number). i also don't think it makes much sense to allow aggregating on numbers (but could have its use cases)

ppisljar commented:
ehm ... under the changes i see this `renamed optimize/.empty → src/manifest.json` ... whats this all about ?

ppisljar commented on src/core_plugins/tag_cloud/public/cloud_vis_params.html:
```diff
@@ -0,0 +1,28 @@
+    <select ng-model="vis.params.textScale" class="form-control">
+      <option value="linear" selected="selected">linear</option>
+      <option value="log">log</option>
+      <option value="sqrt">square root</option>
```
I think options should be defined in cloud.js. Other visualizations do the same (check kbn_vislib_vis_types (map specifically uses different labels/values)) and I think its cleaner than defining them in template files.

ppisljar commented on src/core_plugins/tag_cloud/public/cloud_vis_params.html:
```diff
@@ -0,0 +1,28 @@
+    <select ng-model="vis.params.orientations" class="form-control">
+      <option value="single" selected="selected">single</option>
+      <option value="rightAngled">right angles</option>
+      <option value="multi">multiple</option>
```
same here

ppisljar commented on src/core_plugins/tag_cloud/public/tag_cloud.js:
```diff
@@ -0,0 +1,337 @@
+
+    super();
+
+    this._element = domNode;
```
i don't think our style guide suggests naming local variables with leading underscore.

thomasneirynck replied:
What does the style-guide recommend instead?

ppisljar commented on src/core_plugins/tag_cloud/public/cloud_controller.js:
```diff
@@ -0,0 +1,85 @@
+    });
+
+    tagCloud.setData(tags);
+    await tagCloud.whenRendered();
```
i would prefer a normal promise syntax here `tagCloud.whenRendered().then(updateState);` than the async/await combination.

ppisljar commented on src/core_plugins/tag_cloud/public/tag_cloud.js:
```diff
@@ -0,0 +1,337 @@
+    return this._allInViewBox ? TagCloud.STATUS.COMPLETE : TagCloud.STATUS.INCOMPLETE;
+  }
+
+  async whenRendered() {
```
as TagCloud already extends EventEmitter and you use tagCloud.on('select') to handle clicking/selecting a tag, what do you think about using the same functionality to notify about renderComplete as well ? I see a benefit of using a promise in case when you are not interested in future re-renderings, but this is a bit hard for me to read. Maybe changing the name of whenRendered to something else would help (don't have any good suggestions .. getRenderCompletePromise might be a bit long :)  )

thomasneirynck replied:
Yeah, agreed, it looks weird :$

jbudz commented:
The UI is locking up for me and I'm getting browser errors:
![image](https://cloud.githubusercontent.com/assets/3143860/20391876/f2b61a4e-ac9a-11e6-9859-53e0f89ae2d0.png)

jbudz commented:
Should the custom label be prefixed with the metric agg still?  It looks like in metric/line/bar visualizations it's replaced.

jbudz commented:
Looking really close, one other issue I found is if there's not enough room for the tag I can see it animate off screen.  It doesn't seem to shrink down to a smaller font size when the minimum allows it.  To reproduce, filter down to one phrase and shrink the page until it dissapears.

jbudz commented:
A future enhancement request that may be useful is to scale font size with window side (in addition to score)

thomasneirynck commented:
@jbudz @ppisljar I squashed the PR after rebasing. I'll be working on adding some more tests, but was hoping to get a sign-off on functionality as soon as possible (if you want to refer  to older versions, you can always see here: https://github.com/thomasneirynck/kibana/tree/tagcloud-backup).

There's a few outstanding improvements based on your review. We discussed these in person, and will postpone for first release. I will create separate tracker tickets in the Kibana repo for these once this is merged.
- Excluding non-string fields from the Term aggregation. Clouds with numbers look somewhat "weird", so it would be nice if we could avoid them. Doing this is a larger refactoring issue and requires more changes to the vislib internals
- Auto-shrink words. Now when you shrink the window,  words may disappear from the view. This is because the word no longer fits in the container. We should consider automatically adjusting the min-max font-size range so that words would fit again.
- adding highlighting on hover

tbragin commented:
@thomasneirynck Overall, looking really good! 

A couple of comments:

* Multiple orientations sometimes turn strings upside down -- that seems a little hard to read. If this orientation type is causing a lot of complexity, do we need in in phase 1?
<img width="1437" alt="screen shot 2016-11-18 at 3 46 48 pm" src="https://cloud.githubusercontent.com/assets/933397/20450573/5a73acbc-ada6-11e6-9344-2ed783520e69.png">

* I'm not sure what it means to have a maximum font size less than the minimum font size - feels like we should not allow that. It is also prone to fix fonts such that some elements are thrown out of the visualization, which can be confusing. Do we need this control in phase 1?
<img width="1431" alt="screen shot 2016-11-18 at 3 48 20 pm" src="https://cloud.githubusercontent.com/assets/933397/20450636/bea9a286-ada6-11e6-8bdc-9c22905f3cd2.png">

* Would it make sense to omit the title at the bottom or make it optional? Users already have to name their visualization and the name always shows on the dashboard, so it's possible we'll be showing the same information twice.
<img width="938" alt="screen shot 2016-11-18 at 3 58 06 pm" src="https://cloud.githubusercontent.com/assets/933397/20450779/ec005508-ada7-11e6-8b36-537bb8d1a3cf.png">

thomasneirynck commented:
@tbragin a few quick comments:
- multiple orientations: I can make it so they won't get turned upside down. I'll let you evaluate; if it doesn't come out right, we can take out the option
- min/max font-size: the "auto"-sizing would actually be not all that straightforward. @jbudz had asked for similar improvements. Right now, it autocorrects in the background (so if min is larger than max, max becomes min). I'll look at a better UI-element to replace the dual sliders.
- the title at the bottom, I kind of like that, because it gives immediate context of what you are looking at. But @jbudz also had some reservations about showing it. What if we make it so the user can hide it (or have it be hidden by default)? But I'm happy to take it out ;)
- bad dark theme visual: that's a bug. I'll fix it.

tbragin commented:
@thomasneirynck 

> I can make it so they won't get turned upside down. 

++ If you can make it work, and you don't see the code as too complex, I'm fine keeping it. My point was more around the fact that if this option was causing you trouble, I'd be ok not having it :)

> I'll look at a better UI-element to replace the dual sliders.

++ Makes sense, thanks for the clarification.

>> What if we make it so the user can hide it (or have it be hidden by default)?

I'm ok keeping it if there is an option to hide it.

ppisljar commented on src/core_plugins/tagcloud/public/tag_cloud.js:
```diff
@@ -10,7 +10,8 @@ const ORIENTATIONS = {
   'multiple': (tag) => {
-    return (~~(hashCode(tag.text) % 12) * 15) - 90;
+    const hashcode = Math.abs(hashCode(tag.text));
+    return ((hashcode % 12) * 15) - 90;
```
could you extract all magic numbers to variables to make this more readable ?

ppisljar commented on src/core_plugins/tagcloud/public/tag_cloud.js:
```diff
@@ -0,0 +1,319 @@
+  },
+  'multiple': (tag) => {
+    const hashcode = Math.abs(hashCode(tag.text));
+    return ((hashcode % 12) * 15) - 90;
```
could you extract all magic numbers to variables ?

thomasneirynck commented:
jenkins, test this

thomasneirynck commented:
for reference; using new double-ended slider for font-size selection:

![image](https://cloud.githubusercontent.com/assets/1833023/20490518/423c247a-afdc-11e6-9e58-b4aa9d188002.png)

jbudz commented:
Thoughts on adding a margin?
![image](https://cloud.githubusercontent.com/assets/3143860/20490871/0bdbeb4c-afd5-11e6-811d-6c35d6b875a7.png)

thomasneirynck commented:
@jbudz ow that looks bad. yes, will add margin

thomasneirynck commented:
jenkins, test this

ppisljar commented:
@cjcenizal do you want to take a look at the slider ?

ppisljar commented:
LGTM
